In [1]:
# install libraries
%pip install category_encoders==2.5.0
%pip install hyperopt==0.2.7
%pip install lightgbm==3.3.0
%pip install matplotlib==3.5.1
%pip install numpy==1.22.4
%pip install pandas==1.3.0
%pip install scikit_learn==1.1.2
%pip install seaborn==0.12.0
%pip install shap==0.40.0

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.

  Using cached matplotlib-3.5.1.tar.gz (35.3 MB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Running setup.py clean for matplotlib
Failed to build matplotlib


  error: subprocess-exited-with-error
  
  python setup.py bdist_wheel did not run successfully.
  exit code: 1
  
  [600 lines of output]
  
  Edit mplsetup.cfg to change the build options; suppress output with --quiet.
  
  BUILDING MATPLOTLIB
        python: yes [3.11.5 | packaged by Anaconda, Inc. | (main, Sep 11 2023,
                    13:26:23) [MSC v.1916 64 bit (AMD64)]]
      platform: yes [win32]
         tests: no  [skipping due to configuration]
        macosx: no  [Mac OS-X only]
  
  C:\Users\tramn\anaconda3\Lib\site-packages\setuptools\__init__.py:80: _DeprecatedInstaller: setuptools.installer and fetch_build_eggs are deprecated.
  !!
  
          ********************************************************************************
          Requirements should be satisfied by a PEP 517 installer.
          If you are using pip, you can try `pip install --use-pep517`.
          ********************************************************************************
  
  !!
    dist

  Using cached numpy-1.22.4-cp311-cp311-win_amd64.whl
  Attempting uninstall: numpy
    Found existing installation: numpy 1.24.4
    Uninstalling numpy-1.24.4:
      Successfully uninstalled numpy-1.24.4
Note: you may need to restart the kernel to use updated packages.


ERROR: Could not install packages due to an OSError: [WinError 5] Access is denied: 'C:\\Users\\tramn\\anaconda3\\Lib\\site-packages\\~-mpy\\.libs\\libopenblas64__v0.3.21-gcc_10_3_0.dll'
Consider using the `--user` option or check the permissions.



  Using cached pandas-1.3.0.tar.gz (4.7 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'error'
Note: you may need to restart the kernel to use updated packages.


  error: subprocess-exited-with-error
  
  pip subprocess to install build dependencies did not run successfully.
  exit code: 1
  
  [322 lines of output]
  Ignoring numpy: markers 'python_version == "3.7" and (platform_machine != "arm64" or platform_system != "Darwin") and platform_machine != "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.8" and (platform_machine != "arm64" or platform_system != "Darwin") and platform_machine != "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.7" and platform_machine == "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.8" and platform_machine == "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.8" and platform_machine == "arm64" and platform_system == "Darwin"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.9" and platform_machine == "arm64" and platform_system == "Da

  Using cached scikit-learn-1.1.2.tar.gz (7.0 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'
Note: you may need to restart the kernel to use updated packages.


  error: subprocess-exited-with-error
  
  Preparing metadata (pyproject.toml) did not run successfully.
  exit code: 1
  
  [2264 lines of output]
  Partial import of sklearn during the build process.
  setup.py:128: DeprecationWarning:
  
    `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
    of the deprecation of `distutils` itself. It will be removed for
    Python >= 3.12. For older Python versions it will remain present.
    It is recommended to use `setuptools < 60.0` for those Python versions.
    For more details, see:
      https://numpy.org/devdocs/reference/distutils_status_migration.html
  
  
    from numpy.distutils.command.build_ext import build_ext  # noqa
  INFO: No module named 'numpy.distutils._msvccompiler' in numpy.distutils; trying from distutils
  INFO: C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\HostX86\x64\cl.exe /c /nologo /O2 /W3 /GL /DNDEBUG /MD -IC:\Program Files\Microsoft Visual Studio\2022\Commun

  Obtaining dependency information for numpy>=1.17 from https://files.pythonhosted.org/packages/3f/6b/5610004206cf7f8e7ad91c5a85a8c71b2f2f8051a0c0c4d5916b76d6cbb2/numpy-1.26.4-cp311-cp311-win_amd64.whl.metadata
  Using cached numpy-1.26.4-cp311-cp311-win_amd64.whl.metadata (61 kB)
Using cached numpy-1.26.4-cp311-cp311-win_amd64.whl (15.8 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.22.4
    Uninstalling numpy-1.22.4:
      Successfully uninstalled numpy-1.22.4
Note: you may need to restart the kernel to use updated packages.


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gensim 4.3.0 requires FuzzyTM>=0.4.0, which is not installed.
tables 3.8.0 requires blosc2~=2.0.0, which is not installed.
datasets 2.12.0 requires dill<0.3.7,>=0.3.0, but you have dill 0.3.8 which is incompatible.
numba 0.57.1 requires numpy<1.25,>=1.21, but you have numpy 1.26.4 which is incompatible.


  Obtaining dependency information for numpy from https://files.pythonhosted.org/packages/d8/ec/ebef2f7d7c28503f958f0f8b992e7ce606fb74f9e891199329d5f5f87404/numpy-1.24.4-cp311-cp311-win_amd64.whl.metadata
  Using cached numpy-1.24.4-cp311-cp311-win_amd64.whl.metadata (5.6 kB)
Using cached numpy-1.24.4-cp311-cp311-win_amd64.whl (14.8 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.26.4
    Uninstalling numpy-1.26.4:
      Successfully uninstalled numpy-1.26.4
Note: you may need to restart the kernel to use updated packages.


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gensim 4.3.0 requires FuzzyTM>=0.4.0, which is not installed.
tables 3.8.0 requires blosc2~=2.0.0, which is not installed.
datasets 2.12.0 requires dill<0.3.7,>=0.3.0, but you have dill 0.3.8 which is incompatible.


In [2]:
import warnings
warnings.filterwarnings("ignore")
import os
path = os.getcwd()
import sys
sys.path.append("")
from utils import *

import pandas as pd
import numpy as np

import random

import category_encoders as ce
import lightgbm as lgb
from lightgbm import LGBMClassifier

from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier

from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve
from sklearn.metrics import auc
from sklearn.metrics import precision_recall_curve
from sklearn.metrics import precision_score
from sklearn.metrics import average_precision_score
import bisect

from sklearn.preprocessing import OneHotEncoder

from hyperopt import fmin, tpe, hp, anneal, Trials

import os
import gc

import datetime

import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.float_format', lambda x: '%.3f' % x)
%load_ext autoreload
%autoreload 2

# Description of Columns
| Columns            | Description                                           |
|--------------------|-------------------------------------------------------|
| User_id            | User Identifier                                       |
| employment_type    | Employment Type (Salaried, Self-Employed)             |
| tier_of_employment | Tier of Employment (According to company performance) |
| industry           | Industry of Employer                                  |
| role               | Role of User in company                               |
| work_experience    | Total Work Experience                                 |
| total_income       | Total Income of customer                              |
| gender             | Gender of Customer                                    |
| married            | Single / Married / Divorced / Widowed etc.            |
| dependents         | No. of dependents on user                             |
| home_type          | Rented / Owned etc.                                   |
| pincode            | Middle 3 digit of Pincode                             |
| has_social_profile | Has FB/Twitter or any other social profile            |
| is_verified        | Is Social Profile Verified                            |
| delinq_2yrs        | Number of loan delinquencies in last 2 years          |
| total_payement     | Total Payment made towards loan in last 2 years       |
| received_principal | Total Principal paid towards loan in last 2 years     |
| interest_received  | Total Interest Received towards loan in last 2 years  |
| number_of_loans    | Number of Loans in last 2 years                       |
| emi_1_dpd          | DPD in repayment of 1st EMI (0, 30+, 60+, 90+)        |
| emi_2_dpd          | DPD in repayment of 2nd EMI (0, 30+, 60+, 90+)        |
| emi_3_dpd          | DPD in repayment of 3rd EMI (0, 30+, 60+, 90+)        |
| emi_4_dpd          | DPD in repayment of 4th EMI (0, 30+, 60+, 90+)        |
| emi_5_dpd          | DPD in repayment of 5th EMI (0, 30+, 60+, 90+)        |
| emi_6_dpd          | DPD in repayment of 6th EMI (0, 30+, 60+, 90+)        |
| max_dpd            | Maximum DPD (0, 30+, 60+, 90+) attained               |
| yearmo             | Year and month of loan application (YYYYMM format)    |

## Action Items
- Drop Gender Column (can't use gender as distinguishing factor for credit risk)
- According to yearmo, split data in Train, Validation, Hold Out
- DPD Roll Rate Analysis and Window Roll Rate Analysis for defining label (default = dpd X in first Y months, Define X & Y)

In [3]:
#Defining Parameters
data_path = 'credit_risk_data.csv'# Path of data
#Non feature Columns (not to be used as features in model)
id_cols = ['User_id','emi_1_dpd', 'emi_2_dpd', 'emi_3_dpd', 'emi_4_dpd', 'emi_5_dpd', 'emi_6_dpd', 'max_dpd', 'yearmo']
drop_columns = ['gender']

In [ ]:
# Numeric, Categorical Features, Feature wise null counts and Zero Variance Features
understand_data(df, id_cols)

In [5]:
# Read Data and drop Non usable Columns
df = process_data(data_path, drop_columns)
print(df.drop(columns = id_cols).shape)
df.head(10)

(143727, 17)


User_id   employment_type tier_of_employment  \
0  7013527          Salaried                  B   
1  7014291  Self - Employeed                  D   
2  7014327               NaN                NaN   
3  7014304               NaN                NaN   
4  7031995               NaN                NaN   
5  6541359               NaN                NaN   
6  7033985               NaN                NaN   
7  7014312          Salaried                  B   
8  3299703               NaN                NaN   
9  7014271               NaN                NaN   

                                            industry  \
0  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
1  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
2  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
3  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
4  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
5  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
6  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
7  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiFcO7...   
8  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
9  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   

                               role work_experience  total_income married  \
0  KHMbckjadbckIFGAZSEWdkcndwkcnCCM             1-2    125000.000     Yes   
1  KHMbckjadbckIFGNCSEWdkcndwkcnCCM             10+     61000.000      No   
2  KHMbckjadbckIFGNYSEWdkcndwkcnCCM            5-10    100000.000     NaN   
3  KHMbckjadbckIFGCASEWdkcndwkcnCCM             2-3     30000.000     NaN   
4  KHMbckjadbckIFGCASEWdkcndwkcnCCM             10+     65000.000     NaN   
5  KHMbckjadbckIFGPASEWdkcndwkcnCCM             2-3    215000.000      No   
6  KHMbckjadbckIFGCASEWdkcndwkcnCCM            5-10     90000.000     NaN   
7  KHMbckjadbckIFGOHSEWdkcndwkcnCCM             1-2    125000.000      No   
8  KHMbckjadbckIFGCASEWdkcndwkcnCCM            5-10     50000.000      No   
9  KHMbckjadbckIFGCOSEWdkcndwkcnCCM             10+    110000.000     Yes   

   dependents home_type  ... interest_received number_of_loans emi_1_dpd  \
0           4      rent  ...           852.690               0         0   
1           1  mortgage  ...          4912.530               0         0   
2           3       own  ...          3310.680               0        90   
3           1      rent  ...          1459.760               0         0   
4           3      rent  ...          1793.000               0         0   
5           0  mortgage  ...          4882.360               0         0   
6           4  mortgage  ...          1733.990               0         0   
7           0  mortgage  ...          5361.920               0         0   
8           1      rent  ...          2476.340               0         0   
9           3  mortgage  ...          2234.430               0         0   

   emi_2_dpd  emi_3_dpd  emi_4_dpd  emi_5_dpd  emi_6_dpd  max_dpd  yearmo  
0         90         90         90         90         90       90  202203  
1          0          0          0          0          0        0  202203  
2         90         90         90         90         90       90  202205  
3          0          0          0          0          0        0  202203  
4          0          0          0          0          0        0  202203  
5          0          0          0         30          0       30  202205  
6          0          0          0          0          0        0  202201  
7          0          0          0          0          0        0  202203  
8         30          0          0          0          0       30  202202  
9          0          0          0          0          0        0  202202  

[10 rows x 26 columns]

In [6]:
# Since Data is on timeline
df.yearmo.value_counts()

yearmo
202203    28750
202201    28750
202202    28750
202204    28750
202205    28727
Name: count, dtype: int64

Split data on timeline to make sure model developed is able to predict well on Out of time data
<br>**Train** - Yearmo 202201 to 202203
<br>**Validation** - Yearmo 202204
<br>**Hold Out** - Yearmo 202205 (Out of time data)

In [7]:
train, val, hold_out = data_split(df)

# Define Label (Roll Rate Analysis)
## DPD Roll Rate

In [8]:
# % of customer achieved X dpd (Since everyone starts at 0 dpd so 100% will be at 0 dpd)
dpd_roll_rate(train)

dpd  user_count user_percent
0    0       86250      100.0 %
1   30       22073      25.59 %
2   60        8099       9.39 %
3   90        7926       9.19 %

As we can see
<br>- ~25.6% customer paid back after crossing dpd30
<br>- Most of the people who crossed dpd60 are also crossing dpd 90
<br>**Conclusion** - *DPD60 should be right choice as Recovery rate is low post that*

## window Roll Rate

In [ ]:
# For DPD 60 analyzing % of customer attain dpd60 on various lifecycle of EMI
window_roll_rate(train, 60)

As we can see
<br>- 80.8% of Defaulters dafault in 1st EMI and 5.2% default in 3rd EMI, Post 3rd EMI rate is low, so we can conclude people Most of the Defaulter default in 1st 3 EMIs
<br>**Conclusion** - We can use DPD60 in first 3 months as our label

## Label Creation

In [ ]:
# We will call Class 1 a defaulter
train = create_label(train, dpd = 60, months = 3)
val = create_label(val, dpd = 60, months = 3)
hold_out = create_label(hold_out, dpd = 60, months = 3)

### Label Distribution

In [ ]:
label_distribution([train, val, hold_out], ['Training', 'Validation', 'Hold Out'], 'label')

- Labels are imbalanced towards class 0
- Same level of Imbalance occure across Train, Validation, Hold Out

# Create Features

In [ ]:
# We are adding few features
## - % Amount Paid as interest in past Loan Repayment
## - % of Loans defaulted in last 2 years
train = derived_features(train)
val = derived_features(val)
hold_out = derived_features(hold_out)

# EDA
## Univariate

In [ ]:
id_cols.append('label')
e = eda(train, id_cols)

In [ ]:
# Univariate Distribution of Numeric Features
e.numeric_summary()

In [ ]:
# Categorical Features Summary
e.categorical_summary()

## Bivariate

### Correlation

In [ ]:
e.correlation_plot()

Some Observations
<br> - Total payment and Received Principal are highly correlated (0.96)
<br> - Interest Received and Total payment are mildely correlated (0.67)

### Box Plot vs Label

In [ ]:
# Box Plot Seggregated by Defaulter and Non Defaulters
e.box_plot(group = ['label'])

Observation
- Distribution of Total Payment, Received Principal, Interest Received Ratio is different for Defaulters and Non-Defaulters
<br> 
<br>these can be some important features

# Categorical Features Transformation

Please Refer
<br> [Paper](https://arxiv.org/pdf/1904.13001.pdf)
<br> [Library](https://contrib.scikit-learn.org/category_encoders/)
<br> to understand Target Encoding better

In [ ]:
cat_cols = train.drop(columns = id_cols).select_dtypes(include=['category', 'object']).columns
params = {"verbose":0,
          "cols":None,
          "drop_invariant":False,
          "return_df":True,
          "handle_missing":'value',
          "handle_unknown":'value',
          "min_samples_leaf":5000,
          "smoothing":1}
target_encoder = categorical_encoding(params)
target_encoder.fit(train, cat_cols, 'label')

In [ ]:
train = target_encoder.transform(train)
val = target_encoder.transform(val)
hold_out = target_encoder.transform(hold_out)

# Feature Selection
<br> There can be many methods to do feature selection like Mrmr, KS, Information Gain etc., Random Forest and Deep Decision Tree can also be used as Feature Selection</br>
<br> 
**Random Forest** - Since Random Forest subset features too, so every feature gets chance to split tree
<br> 
**Decision Tree** - Decision Tree Uses Greedy Algorithm to split Tree so choose between 2 correlated features which one is best
<br>
<br>Drop features which are of 0 importance in Random Forest and Decision Tree

## Random Forest

In [ ]:
rf_params = {"n_estimators":250, 'criterion':'entropy','verbose':False, 'n_jobs':25}
rf_zero_imp = random_forest_zero_importance(train, id_cols, 'label', rf_params)

## Decision Tree

In [ ]:
dt_params = {}
dt_zero_imp = decision_tree_zero_importance(train, id_cols, 'label', dt_params)

## Zero Importance Columns

In [ ]:
drop_cols = list(set(rf_zero_imp) & set(dt_zero_imp))
print(drop_cols)

Since Pincode is of 0 importanc in Random Forest and Decision Tree, we will drop this features in model

# Model Development
## Hyperparameter Tuning

We will try to search hyperparameter using Hyperopt, Following are relevant links to understand Hyperopt better
- [Paper](https://conference.scipy.org/proceedings/scipy2013/pdfs/bergstra_hyperopt.pdf)
- [Library](http://hyperopt.github.io/hyperopt/)

Hyperopt tries to minimize of score (to be defined by user) by smartly iterating though various combination of hypeparameters in given space
<br>
Score = (Train ROC AUC - Val ROC AUC + 1)/(Val ROC AUC)
<br>
<br>
**Intution**
- Train ROC AUC - Val ROC AUC will reduce overfitting
<br> 
- 1/(Val ROC AUC) will increase ROC AUC for Validation Set
</br>
After Every Iteration we are appending results to the results of existing runs and saving to hyperopt_results.csv, so in case we halt the run in between, we will not loose results

<br></br>
**Setting**
<br>- Loss Function = LogLoss
<br>- Implementation = Lightgbm Implementation of Boosting Algorithm 
<br>- Primary Evalution Metric - ROC AUC
<br>- Hypeparameter Selection - Hyperopt

In [ ]:
result_ho = pd.DataFrame()
i = 0

## Defining Objective Function
## will return the score to optimise
def objective(space):
    
    global i
    global result_ho
    
    #Creating Lightgbm DataFrame
    lgb_train = lgb.Dataset(train.drop(columns = id_cols+drop_cols), label = train.label)
    lgb_val = lgb.Dataset(val.drop(columns = id_cols+drop_cols), label = val.label)

    #Parameters of model
    params = {
        'num_leaves': int(space['num_leaves']),
        'max_depth': int(space['max_depth']),
        'learning_rate': space['learning_rate'],
        'objective': 'binary',
        'metric': 'auc',
        "boosting": "gbdt",
        'feature_fraction' : space['feature_fraction'],
        'max_bin' : int(space['max_bin']),
        'min_data_in_leaf': int(space['min_data_in_leaf']),
        "min_data_in_bin": int(space['min_data_in_bin']),
        "bagging_freq": 20,
        "random_seed": 2019,
        "lambda_l1": space['lambda_l1'],
        "lambda_l2": space['lambda_l2'],
        'pos_bagging_fraction' : space['pos_bagging_fraction'],
        'neg_bagging_fraction' : space['neg_bagging_fraction'],
        'verbose': -1
    }

    evals_result = {}
    
    clf = lgb.train(params, lgb_train, 20000, valid_sets=lgb_val,
                valid_names='val',
                early_stopping_rounds=50,
                verbose_eval=False, evals_result=evals_result)
    gc.collect()
    
    result = pd.DataFrame(clf.params, index=[0])
    
    ## Calculating AUC
    pred_train = clf.predict(train[clf.feature_name()])
    pred_val = clf.predict(val[clf.feature_name()])
    pred_hold_out = clf.predict(hold_out[clf.feature_name()])

    gc.collect()
    train_auc = roc_auc_score(train.label, pred_train)#, num_iteration=clf.best_iteration)
    val_auc = roc_auc_score(val.label, pred_val)#, num_iteration=clf.best_iteration)
    hold_out_auc = roc_auc_score(hold_out.label, pred_hold_out)#, num_iteration=clf.best_iteration)
    gc.collect()

    score = (abs(train_auc - val_auc) + 1)/((1+val_auc)*(1+val_auc))
    
    result["train_auc"] = train_auc
    result["val_auc"] = val_auc
    result["hold_out_auc"] = hold_out_auc
    result["train_test_diff"] = train_auc - val_auc
    result["n_estimators"] = clf.best_iteration
    result["score"] = score
    
    del clf
    
    result_ho = result_ho.append(result)
    result_ho.to_csv('hyperopt_results.csv', index=False)
    i = i+1
    
    return (score)

In [ ]:
## Hyperparameter space
## Space is selection of data point from the given distribution
## Distribution is defined for every hyperparameter seperately
space = {
    'num_leaves': hp.quniform('num_leaves', 2, 24, 1), # Uniform integer between 2 and 24
    'max_depth': hp.quniform('max_depth', 2, 12, 1), # Uniform integer between 2 and 12
    'learning_rate': hp.uniform('learning_rate', 0.005, 0.015), # Values between 0.005 to 0.015
    'feature_fraction' : hp.uniform('feature_fraction', 0.1, 1), # Values between 0.1 to 1
    'max_bin' : hp.quniform('max_bin', 10, 100, 10),
    'min_data_in_leaf' : hp.quniform('min_data_in_leaf', 25, 1000, 25),
    'lambda_l1' : hp.uniform('lambda_l1', 0, 50),
    'lambda_l2' : hp.uniform('lambda_l2', 0, 50),
    'min_data_in_bin' : hp.quniform('min_data_in_bin', 5, 100, 5),
    'pos_bagging_fraction' : hp.uniform('pos_bagging_fraction', 0.1, 1),
    'neg_bagging_fraction' : hp.uniform('neg_bagging_fraction', 0.1, 1)
    }

In [ ]:
best=fmin(fn = objective, # function to optimize
          space = space, # space from which hyperparameter to be choosen
          algo = tpe.suggest, # optimization algorithm, hyperopt will select its parameters automatically
          max_evals = 50,
          rstate = np.random.default_rng(7)
         )

## Model Selection

From Hyperopt Results we will choose best parameters according to following criteria. It is not necessary we use that particular hyperparameter
- High Score
- High Validation AUC
- Low Overfitting

In [ ]:
hyperopt_results = pd.read_csv('hyperopt_results.csv')

In [ ]:
hyperopt_results.head(5)

In [ ]:
hyperopt_results.sort_values(by = 'score').head(5)

In [ ]:
hyperopt_results.sort_values(by = 'val_auc', ascending=False).head(5)

In [ ]:
# check the minimum score and select the index as best parameter index

best_param_index = hyperopt_results.index[hyperopt_results['score'] == hyperopt_results['score'].min()].tolist()[0]
lgbm_params = dict(hyperopt_results.iloc[best_param_index,:19])

## Selected Model

Lightgbm Details
- [Paper](https://proceedings.neurips.cc/paper/2017/file/6449f44a102fde848669bdd9eb6b76fa-Paper.pdf)
- [Library](https://lightgbm.readthedocs.io/en/v3.3.2/)

In [ ]:
lgb_train = lgb.Dataset(train.drop(columns = id_cols+drop_cols), label = train['label'])
lgb_val = lgb.Dataset(val.drop(columns = id_cols+drop_cols), label = val['label'])
evals_result = {}
clf = lgb.train(lgbm_params, lgb_train, 20000, valid_sets=lgb_val,
                valid_names='val',
                early_stopping_rounds=50,
                verbose_eval=False, evals_result=evals_result)

# Model Evalution

In [ ]:
#Predictions from Model on Training, Validation and Hold Out
pred_train = clf.predict(train[clf.feature_name()])
pred_val = clf.predict(val[clf.feature_name()])
pred_hold_out = clf.predict(hold_out[clf.feature_name()])

### ROC AUC, PR AUC, KS Stats

In [ ]:
#Printing ROC AUC, PR AUC for Class 1 and Class 0
print("ROC AUC")
roc_auc([train.label,val.label,hold_out.label], [pred_train, pred_val, pred_hold_out])
print("")
print("Class 1 PR AUC")
pr_auc([train.label,val.label,hold_out.label], [pred_train, pred_val, pred_hold_out])
print("")
print("Class 0 PR AUC")
pr_auc([1-train.label,1-val.label,1-hold_out.label], [1-pred_train, 1-pred_val, 1-pred_hold_out])
print("")

### ROC Curve

In [ ]:
roc_auc_curve([train.label,val.label,hold_out.label], [pred_train, pred_val, pred_hold_out])

### PR Curve

In [ ]:
print("PR Curve for Class 1")
pr_auc_curve([train.label,val.label,hold_out.label], [pred_train, pred_val, pred_hold_out])

In [ ]:
print("PR Curve for Class 0")
pr_auc_curve([1-train.label,1-val.label,1-hold_out.label], [1-pred_train, 1-pred_val, 1-pred_hold_out])

### Score Distribution

In [ ]:
score_distribution([train.label,val.label,hold_out.label], [pred_train, pred_val, pred_hold_out], ["Train","Val","Hold Out"])

## Feature Importance
### Split & Gain

In [ ]:
feature_importance = pd.DataFrame({"feature":clf.feature_name(),"split":clf.feature_importance('split'), "gain":clf.feature_importance('gain')}).sort_values(by = 'gain', ascending = False)
feature_importance.iloc[:10,:]

### SHAP

Literature about SHAP
<br>
- [Paper](https://arxiv.org/pdf/1705.07874.pdf)
- [Library](https://shap.readthedocs.io/en/latest/index.html)
</br>

**How to Read SHAP**
- Value of features on color scale vary from Blue to Red where Red is high and Blue is low
- Right Side of Grey Vertical Line is for class 1 and Left side is for Class 0

In [ ]:
import shap
data_list = [train,val,hold_out]
data_type_list = ['Training', 'Validation', 'Hold Out']
shap_importance(clf, data_list, data_type_list)

Key Points
- Top features are consistent across Training, Validation, Hold Out (Higher the inconsistency, higher the chances the model is overfitting and will decay soon)

# Making Sense of Score
## Class Rate Curve

This Curve Tells 2 important things
1. Default rate across different score bucket for Train, Validation and Hold Out
2. User Distribution across different score bucket for Train, Validation and Hold Out

<br>
In Case User Distribution across different score bucket is very inconsistent in Train, Validation and Hold Out that means there is shift in domain of Out of Time Data

In [ ]:
target_list = [train.label, val.label, hold_out.label]
pred_list = [pred_train, pred_val, pred_hold_out]
data_type_list = ['Train', 'Val', 'Hold Out']
class_rate(target_list, pred_list, data_type_list)

## Right Threshold

In [ ]:
cutoff_score(val.label, pred_val, 0.01)